content/blog/2025/07-01_httr2_client/index.qmd

```diff
@@ -0,0 +1,623 @@
+        clipboard-write=(), hid=(), serial=(), cross-origin-isolated=(), execution-while-not-rendered=(self),
+        execution-while-out-of-viewport=(self), keyboard-map=(), navigation-override=(self),
+        identity-credentials-get=(), idle-detection=(), local-fonts=(), otp-credentials=(),
+        window-management=(), storage-access=()
+      content-security-policy: object-src 'none'; frame-ancestors 'self'; upgrade-insecure-requests
+      tdm-reservation: '1'
+    body:
+      encoding: ''
+      file: no
+      string: '{"isPersonalDataResponsible":false,"displayName":"ccda25ce-8256-4c6f-ba71-7a4357dc6caf@apiclient","logoutLink":"/signout","isSuperUser":false,"isAuthenticated":true,"userType":"UNKNOWN_ROLE","hasAcceptedTos":true,"isSupportUser":false,"isAdministrativeUser":true,"isInLdapGroupUioTils":false}'
+  recorded_at: 2025-06-06 12:11:47 GMT
+  recorded_with: vcr/1.6.0, webmockr/2.0.0
+```
+
+There is a lot of information here, about the request made, the status code of the response, and the response it self.
+This is great stuff!
+Every time I ran the test, it was passing and I was happy.
+I committed my code and sent it off to GitHub, where all my package checking actions subsequently failed.
+Which confused me, since I thought given I recorded the responses with vcr, that wouldn't happen?
+
+Where I got my logic wrong, was that what was failing wasn't the API calls (or vcr's accessing of the cassettes), but rather my own checking function for whether authentication was set up (which it was not on GitHub Actions!)
+
+Remember I made a `ns_has_auth` function, that checks if authentication is set up?
+Well, that's integrated in the `ns_client` function, meaning _before_ any calls to the API!
+So my GitHub Actions. were failing before any cassettes were in play.
+
+This is where the dreaded moment I knew was going to happen in this journey would indeed happen, I needed to figure out how to _mock_ a function.
+Mocking in a test is used to circumvent a specific piece of code from testing.
+This sounds counterintuitive, since we are supposed to be testing the code, but it is very necessary in certain cases.
+
+In this particular case, I don't need to test the `ns_has_auth()` function, that is not the functionality I am after testing.
+I can test that in another test specifically made to test it, where I can control input and results better.
+In this case, I want to test the API call and make sure vcr has recorded a good cassette and that this cassette is used in instances where I need it to (like on GitHub Actions. and during CRAN checks).
+
+With all that being said, I need to make sure that the `ns_has_auth()` function always returns true and can continue on to the remaining function where I ask it to during testing.
+We create a function with `testthat::local_mocked_bindings` which we will wrap around any code where we want to test the other parts of the function rather than checking if authentication is set up.
```

Comment:
```suggestion
We create a function with [`testthat::local_mocked_bindings()`](https://testthat.r-lib.org/reference/local_mocked_bindings.html) which we will wrap around any code where we want to test the other parts of the function rather than checking if authentication is set up.
```